redis7/redis.go

```diff
@@ -0,0 +1,353 @@
+		opts.InvokeOnRollback()
+		return appError
+	}
+
+	return nil
+}
+
+func (s *store) checkOpen() error {
+	s.mux.RLock()
+	defer s.mux.RUnlock()
+	if s.client == nil {
+		return stoabs.ErrStoreIsClosed
+	}
+	return nil
+}
+
+type tx struct {
+	reader redis.Cmdable
+	writer redis.Cmdable
+	store  *store
+}
+
+func (t tx) GetShelfWriter(shelfName string) (stoabs.Writer, error) {
+	return t.store.getShelf(shelfName, t.writer, t.reader), nil
+}
+
+func (t tx) GetShelfReader(shelfName string) (stoabs.Reader, error) {
+	reader, err := t.store.getShelfReader(shelfName)
+	if err != nil {
+		return nil, err
+	}
+	if reader == nil {
+		return nil, nil
+	}
+	return reader, nil
+}
+
+func (t tx) Store() stoabs.KVStore {
+	return t.store
+}
+
+func (t tx) Unwrap() interface{} {
+	return t.writer
+}
+
+type shelf struct {
+	name   string
+	prefix string
+	reader redis.Cmdable
+	// writer holds the Redis command writer for writing to the database.
+	// It is separate from the reader because writes use a transactional Redis pipeline, which buffers the commands.
+	// Buffering read commands would result in unexpected behaviour where the reads are only executed on transaction commit.
+	writer redis.Cmdable
+	store  *store
+}
+
+func (s shelf) Put(key stoabs.Key, value []byte) error {
+	if err := s.store.checkOpen(); err != nil {
+		return err
+	}
+	return s.writer.Set(context.TODO(), s.toRedisKey(key), value, 0).Err()
+}
+
+func (s shelf) Delete(key stoabs.Key) error {
+	return s.writer.Del(context.TODO(), s.toRedisKey(key)).Err()
+}
+
+func (s shelf) Get(key stoabs.Key) ([]byte, error) {
+	result, err := s.reader.Get(context.TODO(), s.toRedisKey(key)).Result()
+	if err == redis.Nil {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	return []byte(result), nil
+}
+
+func (s shelf) Iterate(callback stoabs.CallerFn) error {
+	var cursor uint64
+	var err error
+	var keys []string
+	for {
+		scanCmd := s.reader.Scan(context.TODO(), cursor, s.toRedisKey(stoabs.BytesKey("*")), resultCount)
+		keys, cursor, err = scanCmd.Result()
+		if err != nil {
+			return err
+		}
+		err := s.visitKeys(keys, callback)
+		if err != nil {
+			return err
+		}
+		if cursor == 0 {
+			// Done
+			break
+		}
+	}
+	return nil
+}
+
+func (s shelf) Range(from stoabs.Key, to stoabs.Key, callback stoabs.CallerFn) error {
+	keys := make([]string, 0, resultCount)
+	// Iterate from..to (start inclusive, end exclusive)
+	var numKeys = 0
+	for curr := from; bytes.Compare(curr.Bytes(), to.Bytes()) == -1; curr = curr.Next() {
+		keys = append(keys, s.toRedisKey(curr))
+		// We don't want to perform requests that are really large, so we limit it at resultCount keys
+		if numKeys >= resultCount {
+			err := s.visitKeys(keys, callback)
+			if err != nil {
+				return err
+			}
+			keys = make([]string, 0, resultCount)
+			numKeys = 0
+		} else {
+			numKeys++
+		}
+	}
+	return s.visitKeys(keys, callback)
+}
+
+func (s shelf) visitKeys(keys []string, callback stoabs.CallerFn) error {
```

Comment:
pass a slice of stoabs.Key so you can omit the `fromRedisKey` which won't work since you don't know the keytype.

Reply:
`fromRedisKey` is required because the KVStore interface returns keys (from Redis) to the caller in `Iterate` and `Range`. The keys that the caller passes are in non-qualified (without the shelf prefixed) format. In other words; the keys passed to and from KVStore by the caller is in simple format (without database/shelf prefix), thus we need to parse keys when returning them to the caller. Otherwise the API would be asymmetric (you can't use keys received through `Iterate` and `Range` in `Get`/`Delete` etc).

Now I think of it, how would that work if we can't just guess/derive the key type? If you have a key returned by Range/Iterate and call `Next` on it? (also applies to BBolt)